ENH: support for clip and quantile ops on DoubleColumns

jreback opened the pull request.

jreback commented on ibis/pandas/tests/test_operations.py:
```diff
@@ -904,3 +904,20 @@ def test_cast_to_decimal(t, df, type):
+        (methodcaller('quantile', 0.5), methodcaller('quantile', 0.5)),
+    ]
+)
+def test_arrayfunctions(t, df, ibis_func, pandas_func):
```
@cpcloud where do you tests for exceptions that bubble up (e.g. an out-of-range for quantile) for instance.?

cpcloud replied:
Maybe have two tests here? I know there's a few examples where we dump a bunch into one, but these are distinct enough that it probably makes sense to separate them.

cpcloud replied:
I would just have a separate test that only contains failures (using `with pytest.raises`). Easier to debug that way.

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -349,6 +349,17 @@ def output_type(self):
 
+class Clip(ValueOp):
+    input_type = [number(name='lower', allow_boolean=False, optional=True),
+                  number(name='upper', allow_boolean=False, optional=True)]
```
What does the behavior when neither is passed?

jreback replied:
just returns the original. not great. I could raise if not at least 1 is passed.

cpcloud replied:
Yeah, probably should raise during execution since we don't have a rule to express "at least one argument" that I'm aware of.

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -349,6 +349,17 @@ def output_type(self):
+    input_type = [number(name='lower', allow_boolean=False, optional=True),
+                  number(name='upper', allow_boolean=False, optional=True)]
+
+    def output_type(self):
```
The output type here should be the same as the input type. You can use `rules.type_of_arg(0)` to get the output type of the first input argument. You should also be able to just assign that directly to `output_type` like this:

```python
class FooNode(...):
    output_type = rules.type_of_arg(0)
```

jreback replied:
done

cpcloud commented on ibis/pandas/execution.py:
```diff
@@ -170,6 +170,20 @@ def execute_series_natural_log(op, data, scope=None):
 
 
+@execute_node.register(
+    ops.Clip, pd.Series, (float, int, type(None)), (float, int, type(None))
```
For `int` use `integer_types` which includes numpy integers and `six.integer_types`. If that's not imported you can import it from `ibis/pandas/core.py`.

Also, does `Series.clip` allow `Series` as input? Do we want to allow other table columns as inputs?

jreback replied:
it does, will add

cpcloud commented on ibis/pandas/execution.py:
```diff
@@ -170,6 +170,20 @@ def execute_series_natural_log(op, data, scope=None):
+@execute_node.register(
+    ops.Quantile, pd.Series, float
+)
+def execute_series_quantile(op, data, q, scopy=None):
```
Misspelling here: `s/scopy/scope/`

jreback replied:
done

cpcloud replied:
Should we validate that `0 <= q <= 1` here?

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -696,6 +707,12 @@ class Mean(Reduction):
 
+class Quantile(Reduction):
+
+    input_type = [number(name='q', allow_boolean=False)]
```
Maybe give this a slightly longer name :)

jreback replied:
done

cpcloud commented on ibis/expr/api.py:
```diff
@@ -1016,6 +1016,25 @@ def log(arg, base=None):
+    -------
+    clipped : type depending on input
+      Decimal values: yield decimal
+      Other numeric values: yield integer (int32)
```
I don't think you need to enumerate any types here. Something like "The type of the returned column is the same as type of the input" or something similar.

jreback replied:
sure

jreback commented:
pushed, but still have to handle this

```
In [3]: pd.Series([1, 2, 3]).quantile(0.5)
Out[3]: 2.0

In [4]: pd.Series([1, 2, 3]).quantile([0.5, 0.75])
Out[4]: 
0.50    2.0
0.75    2.5
dtype: float64

```
input type is easy, and output type is a the type of inputscalar column, But we don't have an Index (e.g. to annotate the quantiles). How do you handle this? (return a Table)?

cpcloud commented on ibis/pandas/tests/test_operations.py:
```diff
@@ -904,3 +911,49 @@ def test_cast_to_decimal(t, df, type):
+    ]
+)
+def test_arraylike_functions_transforms(t, df, ibis_func, pandas_func):
+    if isinstance(pandas_func, Exception):
```
Why do you need this? It looks like every `pandas_func` is a lambda.

jreback replied:
changed in folllowing push

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -696,6 +708,14 @@ class Mean(Reduction):
+
+    input_type = [value,
+                  number(name='quantile', allow_boolean=False),
+                  string(name='interpolation', optional=True)]
```
I think we want to use `rules.string_options(list_of_possible_interpolation_values)` so that we can limit these before executing.

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -696,6 +708,14 @@ class Mean(Reduction):
+    input_type = [value,
+                  number(name='quantile', allow_boolean=False),
+                  string(name='interpolation', optional=True)]
+    output_type = rules.scalar_output(_array_reduced_type)
```
This is correct for multiple quantiles, but not for a single quantile. In this case the `output_type` is actually the same as the *second* argument: `T -> T` or `array<T> -> array<T>`. `rules.type_of_arg(1)` should cover this case.

cpcloud replied:
I don't remember how to spell the correct type for the input though, which is "any scalar or array of scalars".

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -696,6 +708,14 @@ class Mean(Reduction):
+class Quantile(Reduction):
+
+    input_type = [value,
+                  number(name='quantile', allow_boolean=False),
```
We don't have a nice way of describing "numeric scalar or array of numeric scalars". I'll put up a PR to do this now.

cpcloud replied:
This is actually pretty tricky because of the distinction between a generic list of values (whose elements could each be of a different type) and an array which is a list of values all with the same type.

I'll make an issue about this to refactor this part of the code.

In the meantime, you can work around this by creating an additional expression type, maybe `MultipleQuantile` and then in the `def quantile` function try to construct one and if that fails validation construct the other.

cpcloud replied:
#1091

jreback commented on ibis/pandas/execution.py:
```diff
@@ -170,6 +170,30 @@ def execute_series_natural_log(op, data, scope=None):
+
+
+@execute_node.register(
+    ops.Quantile, pd.Series, float,
```
I repeated this w/o the optional arg as I couldn' get ``rules.string_options`` to be optional....?

cpcloud replied:
Did you try dispatching on `(float, type(None))`?

cpcloud replied:
You should be able to cover this in one function by dispatching on `six.string_types + (type(None),)` for the `interpolation` parameter.

jreback commented on ibis/expr/operations.py:
```diff
@@ -696,6 +708,17 @@ class Mean(Reduction):
+                  rules.string_options(
+                      ['linear', 'lower', 'higher',
+                       'midpoint', 'nearest'],
+                      name='interpolation', optional=True)]
```
e.g. here I *think* the optional is being ignored

cpcloud replied:
Hm, could be a bug. Looking into it.

cpcloud commented:
@jreback let's handle the array of quantiles case in a follow up. So for now we only allow a scalar double as the quantile argument. There's some other things to fix before we can properly handle the array case.

jreback commented:
sgtm

will fix up the doc string and add some tests

jreback commented:
pushed

jreback commented on docs/source/release.rst:
```diff
@@ -7,6 +7,17 @@ Release Notes
     interesting. Point (minor, e.g. 0.5.1) releases will generally not be found
     here and contain only bug fixes.
 
+0.12.0 (???)
```
leave out? change to 0.11.3?

cpcloud replied:
Yep, this should be 0.11.3

cpcloud commented on docs/source/release.rst:
```diff
@@ -7,6 +7,17 @@ Release Notes
+0.12.0 (???)
+------------
+
+This release brings initial Pandas backend support along with a number of
```
Maybe just leave the section blank for now.

cpcloud commented on ibis/expr/api.py:
```diff
@@ -1016,6 +1016,55 @@ def log(arg, base=None):
+
+    Parameters
+    ----------
+    quantile : float, default 0.5 (50% quantile)
```
Doesn't look like there's a default, so should remove this text.

cpcloud commented on ibis/expr/operations.py:
```diff
@@ -696,6 +708,17 @@ class Mean(Reduction):
+class Quantile(Reduction):
+
+    input_type = [value,
+                  number(name='quantile', allow_boolean=False),
```
This should be `rules.double()`

jreback replied:
technically this should work on all numeric (and datetimelikes) actually

cpcloud replied:
Oh interesting re datetime.

cpcloud commented:
@jreback few more small things. other than that this looks good to go.

jreback commented:
@cpcloud if you'd have a look. I added ``default='linear'`` for Quantile. but still looking for original signature (and now that I have deleted it, the testsfail :>)

jreback commented:
pushed with updates

cpcloud commented:
I'll merge on green

jreback commented:
green now

cpcloud commented:
thanks @jreback !
